


The environment I’m using to run sourmash on the HPC is something like this:

	module load GNU/6.2
	module load Python/2.7.2	

and then create & activate a virtualenv and install the following:
	
	virtualenv dev
	source dev/bin/activate
	
(you can put that list in ‘requirements.txt’ and run ‘pip install -r 
requirements.txt’ and it should work.)
	
	mkdir 2017_sourtax_project
	cd 2017_sourtax_project/

upgrade setuptools and pip 

	pip install --upgrade setuptools
	pip install --upgrade pip

make a text file containing install requirements
 
	nano requirements.txt
	
	
and requirements.txt looks like 
	
	bz2file==0.98
	coverage==4.2
	cycler==0.10.0
	matplotlib==1.5.3
	numpy==1.11.2
	py==1.4.31
	pyparsing==2.1.10
	pytest==3.0.4
	pytest-cov==2.4.0
	python-dateutil==2.6.0
	pytz==2016.7
	PyYAML==3.12
	scipy==0.18.1
	screed==0.9
	
and then install requirements
	
	pip install -r requirements.txt
	
Install khmer

	pip install git+https://github.com/dib-lab/khmer.git@master

	
Then finally get the master branch of sourmash 
	
	https://github.com/brooksph/sourmash.git	
	
and install
	
	cd sourmash 
	make install 
			
copy some files over 

	cp /mnt/research/ged/sherine/2015-metagenome-assembly/pipeline/megahit-quality-assembly.fa .

	cp /mnt/research/ged/sherine/2015-metagenome-assembly/pipeline/mircea.fa
	
	
-----------------------------------------------------------------------------

Error trim some seqs 

	trim-low-abund.py -M 16e9 -k 20 SRR606249.pe.qc.fq.gz

-----------------------------------------------------------------------------

Finding unique k-mers 

	unique-kmers.py --report lindgreen_unique_kmers *interleaved.cat.fq.gz
	
-----------------------------------------------------------------------------	

#### build signatures from mircea reference genomes from shakya et al 
	sourmash compute -k 21 --dna -n 0 --scaled 10000 mircea.fa
	
#### note: the singleton flag should only be used when the input file consists of a set of genomes not a set of reads

#### make SBT (i.e. database)
	sourmash sbt_index -k 21 --dna mircea mircea.fa.sig

#### build a signature from the metagenome (megahit-quality-assembly.fa is the Megahit assembly generated by Sherine)
	sourmash compute -k 21 --dna --scaled 10000 -n 0 megahit-quality-assembly.fa

#### search against mircea
	sourmash sbt_gather -k 21 --dna mircea megahit-quality-assembly.fa.sig --threshold=0.001

------------------------------------------------------------------------------
